### ISSUE TITLE
Enhancement: [no-misused-promises] checksVoidReturn: Allow async function if it never throws

### ISSUE THREAD

**Issue by phaux** [typescript-eslint/typescript-eslint#5543]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-misused-promises#checksvoidreturn

### Description

As the documentation says, passing a promise-returning callback where a void callback is expected can result in exceptions being silently ignored, as they will only show in the console.

To correctly handle the error, one can make the callback function sync and only call the actual async function from it. `no-floating-promises` then reminds us to handle the error of a non-awaited promise:

```ts
async function handleClick() {
  await maybeFail();
  await doSomething();
}

// no error ✅ ok.
// the exception is properly handled. 
document.addEventListener("click", () => {
  handleClick().catch(error => console.error("click failed", error));
});
```

We can also not use async/await at all and limit ourselves to only `then`s:

```ts
// no error ✅ ok.
// the exception is properly handled. 
document.addEventListener("click", () => {
  maybeFail()
    .then(doSomething)
    .catch(error => console.error("click failed", error));
});
```

However there are other methods to handle the exception, but unfortunately `no-misused-promises` doesn't allow them...

### Fail

```typescript
// If we forget to handle the error, 
// either `no-misused-promises` or `no-floating-promises` does remind us about it.

// error: no-floating-promises ✅ ok.
document.addEventListener("click", () => {
  maybeFail();
  doSomething();
});

// error: no-misused-promises ✅ ok.
document.addEventListener("click", async () => {
  await maybeFail();
  await doSomething();
});
```

### Pass

```typescript
// error: no-misused-promises ❌ but should be allowed.
// if the whole async function is wrapped in a try/catch, it should be considered safe.
document.addEventListener("click", async () => {
  try {
    await maybeFail();
    await doSomething();
  }
  catch (error) {
    console.error("click failed", error)
  }
});

// error: no-misused-promises ❌ but should be allowed.
// if we handle the error of returned promise, it should be considered safe.
document.addEventListener("click", () => {
  return maybeFail()
    .then(doSomething)
    .catch(error => console.error("click failed", error));
});

// error: no-misused-promises ❌ but should be allowed.
// if we handle the error of every awaited promise, it should be considered safe.
document.addEventListener("click", async () => {
  await maybeFail().catch(error => console.error("click failed", error));
  await doSomething().catch(error => console.error("click failed", error));
});
```

### Additional Info

[Playground](https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwFsoBPAIxADEosIAKASgC54AFGHArAZxAB4A3HFmAA+AFBiA9JPioc8EDHYx4gUHJ4OANYA6KTIwALBCAAeYEAAdseeN3gX2FxRGLwDUVMAghg2+GOAcMGQCEFQMbShgYABRfjCMABluDDDFWgAiSCwwTQyAGngGeABeEXgAbzF4QhJyKhoGbTAoDDADWkVlUvKwPC4cb20unBhM7Nykam9gAoUlUfp6AG4xAF8ViWl55WY5AFpECBxWrFQAc32HDm4QLjUNHQCgkITI6LiE5K5U1HSsiA5PKFYplSrVWpkSjTBirDarPQ7UZ7HD7ThcZA8YBXdjou4PLS6QLBULhd6xeLhb6-f4TYHwKBcYjoIr0HrgmpQADu1AwkPqMM28K2MhGMBRaO4mJ8OJuPHugBlyeCkZB8rgGHDICDAZUIKAQY5cny6bZYRDwQwILka7wMpkslDoaz4OxcmBQCxOHVnBkWmDESQtNoGQpYNUarU68jwPqoLjCRQ+eBcKCIEBEl6kiJRClfFJpMYAoFzRnMsCs9lVGoYf0cmoMnlh-nQxqbGprCFB9pFMVsqv12MDIZi8aAyaIaY+Oa9iEd4WIsUS9HS7HXPGK5Wq5MR7W6hkGnBG3yIs3wI1uDxeBCWpEqHDmkDxWvc3k+SRwDDIGB-HVr26h8NNV3aNBwTOAdRTNMMxJN4c0+Kl8z+Qs6RLe1y1Bcp+xfJsiChBo6HoZpWnaToFhUMFB0GdMRyLcdJ1mQpe3odZNjEIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRIAGnDY8RUsgrU6DZm060AZvA4BDfMwDmo8ZJToochZFUBfEG6A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

I think it should be the default behavior, but an option would suffice too.

I can send a PR.

**Comment by bradzacher** [comment]
The issue with this is that it's quite difficult to check for a "function that handles all its errors".

For the most part a human can verify that a function looks like it handles its errors, but it's hard for a lint rule.

For example:
```ts
document.addEventListener("click", async (evt) => {
  if (evt.target === something) {
    return;
  }

  try {
    await somethingAsync();
  } catch {}
});
```

This code is obviously safe and a human can easily see that - but it's really hard to teach a lint rule it's safe, which means instead someone has to change their code to this to make it pass linting:
```ts
document.addEventListener("click", async (evt) => {
  try {
    if (evt.target === something) {
      return;
    }

    await somethingAsync();
  } catch {}
});
```

Which is not ideal at all! Also worth noting that the first code might not actually be safe either - imagine if the types are wrong and `evt` is null - now the code someone thought was safe actually isn't and throws an exception!

Another example might be like:
```ts
document.addEventListener("click", async (evt) => {
  try {
    await somethingAsync1();
  } catch {}

  try {
    await somethingAsync2();
  } catch {}
});
```

Again easy for a human, hard for a lint rule.

Or there's cases like:
```ts
catch {
  throw new Error();
}

catch {
  throwsError();
}
```

Beating a dead horse here - but it's *really* hard to validate that a given function never throws!

IMO a better course of action is to do something like use the `void` operator to signify you're not awaiting a function on purpose and accept that it's a liability:

```ts
document.addEventListener("click", (evt) => {
  void (async () => {
    try {
      await somethingAsync();
    } catch {}
  })();
});

// or

async function handler(evt) {
  try {
    await somethingAsync();
  } catch {}
}

document.addEventListener("click", (evt) => {
  // I solemnly swear I know what I'm doing
  void handler(evt);
});
```

**Comment by phaux** [comment]
@bradzacher 

> The issue with this is that it's quite difficult to check for a "function that handles all its errors".

I'd say let's at least only allow the case when the whole body is wrapped in try/catch. It's basically equivalent to async IIFE with `.catch` which is currently allowed. 

> [this] is not ideal at all!

I still think it looks better than an async IIFE. An IIFE adds a little bit of wierdness to the code, because it's not common in other languages, and I think it's good practice to avoid it when possible.



> imagine if the types are wrong and `evt` is null - now the code someone thought was safe actually isn't and throws an exception!

That's also a problem with regular non-async callback functions, and I think it's okay to not check those because TypeScript itself does a pretty good job with finding these errors.

So I would be fine with either a) requiring whole body to be try/catch or b) consider only calls and awaits to be unsafe, because TypeScript doesn't track the throw type of functions and promises (it's always `any`), and require them to be contained in try block.

Also if the types are wrong then many recommended rules will stop working and I think that's okay.

> Or there's cases like:
> 
> ```ts
> catch {
>   throw new Error();
> }
> 
> catch {
>   throwsError();
> }
> ```

That's also the case with `no-floating-promises`:

```ts
// this catch counts as handling the error
Promise.resolve().catch(() => {
  throw new Error();
});
```

**Comment by JoshuaKGoldberg** [comment]
I'm going to go ahead and close this since it's been a while without comment. It's a known issue that `@typescript-eslint/no-floating-promises` can be a little harsh, but there's no practical way for us to ease it up given the constraints mentioned :point_up:. Thanks! ✨
